### ISSUE TITLE
decorator on dataclass does not respect the __init__ signature 

### ISSUE THREAD

**Issue by SylvainGuieu** [microsoft/pyright#7546]
Use case:  a registry of factories,  where the factory can be a callable but also be a type ( dataclassed ) 

```python
from dataclasses import dataclass, field
from typing import Any, Protocol

class GetterProtocol(Protocol):
    def get(self,key:str)->Any:
        ...

@dataclass
class Config:
    """ Some config """
    ...


class FactoryProtocol(Protocol):
    def __call__(self, config: Config)->GetterProtocol:
        ... 


@dataclass 
class Registery:
    registery: list[FactoryProtocol] = field(default_factory=list)
    def register( self, factory: FactoryProtocol)->FactoryProtocol:
        self.registery.append( factory )
        return factory # return factory to be used as decorator 

my_registery = Registery() 


#######################
# First usage  way -> Error 
@my_registery.register
@dataclass
class MyGetter:
    config: Config
    def get(self, key:str)->Any:
        ...
# The above decoration trow an diagnostic error 
# Diagnostics:
# 1. Argument of type "type[MyGetter]" cannot be assigned to parameter "factory" of type "FactoryProtocol" in function "register"
#      Type "type[MyGetter]" cannot be assigned to type "(config: Config) -> GetterProtocol"
#        Function accepts too many positional parameters; expected 0 but received 1 [reportArgumentType]

######################
# Second usage way -> OK 
my_registery.register(MyGetter)
# OK  
```

I would have expected that  the 1st way (decorator) would give the same result than the second. 
PS : Of course this works fine if not using dataclasses

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The way dataclasses work, dataclass fields are allowed to reference the dataclass itself in their type definitions. Any decorator applied to the class must be evaluated before these field types can be evaluated, and these field types must be evaluated before the `__init__` method is synthesized. There's no good way to break this circularity other than to say that the decorator cannot rely on the synthesized method.

If your decorator relies on the dataclass `__init__` method, you'll need to manually define it rather than relying on pyright (or the runtime) to synthesize it for you.

**Comment by SylvainGuieu** [comment]
hum ... 
For me : 
```python
@decorator2
@decorator1
class A:
    ....
```

```python
@decorator1
class A:
    ... 
A = decorator2( A ) 
```

```python
class A:
    ... 
A = decorator2( decorator1(A) ) 
```

Are grammatically strictly equivalent. 

So (In my mind) results should be the same. 

According to what you say, then both way I exposed in original issue should trough the same diagnostic error.

**Comment by erictraut** [comment]
No, the two cases in your example are not equivalent. In the case of `my_registery.register(MyGetter)`, you are not assigning the result back to the `MyGetter` symbol as is done with a decorator.

**Comment by SylvainGuieu** [comment]
Yes of course, but regarding the diagnostic they are equivalent, isn't ? :
 
```python 
MyGetter = my_registery.register(MyGetter)
# OK 
MyGetter2: FactoryProtocol = my_registery.register(MyGetter)
# OK
MyGetter3: type[GetterProtocol] = my_registery.register(MyGetter)
# OK
``` 

They are all okay, as they should.  I do not understand how they are different to : 

```python
@my_registery.register
@dataclass 
class MyGetter:
    ...
```

Is it different in the python compiler ?

**Comment by erictraut** [comment]
The Python runtime doesn't evaluate types at the time it synthesizes the `__init__` method. Types are ignored at runtime, so this circular dependency isn't an issue.

As I said, pyright is working as intended here, so you'll need to change your code if you want it to work with static type checking. This isn't a bug.

**Comment by SylvainGuieu** [comment]
Okay then not a bug but probably a feature or limitation of pyright. 
I mean, there are two syntaxes to write exactly the same thing and there never be anything you could do at python run time to distinguish one to another. Both will work exactly the same way. 

I understand you have your developer logic but as "user" having  two exact same thing with 2 different results in pyright looks more as a feature I have to leave with than something "normal".

cheers.
